File: app/src/main/java/org/ametro/utils/ListUtils.java
package org.ametro.utils;

import java.util.ArrayList;
import java.util.Collection;

public class ListUtils {

    public static <T> T firstOrDefault(Collection<T> target, T defaultValue) {
        if(target == null || target.size() == 0){
            return defaultValue;
        }
        return target.iterator().next();
    }

    public interface IPredicate<T> { boolean apply(T type); }

    public static <T> Collection<T> filter(Collection<T> target, IPredicate<T> predicate) {
        Collection<T> result = new ArrayList<>();
        for (T element: target) {
            if (predicate.apply(element)) {
                result.add(element);
            }
        }
        return result;
    }
}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/utils/ListUtils.kt
package io.github.romangolovanov.apps.ametro.utils

object ListUtils {

    @JvmStatic
    fun <T> firstOrDefault(target: Collection<T>?, defaultValue: T): T {
        if (target == null || target.isEmpty()) {
            return defaultValue
        }
        return target.iterator().next()
    }

    fun interface IPredicate<T> {
        fun apply(type: T): Boolean
    }

    @JvmStatic
    fun <T> filter(target: Collection<T>, predicate: IPredicate<T>): Collection<T> {
        val result = ArrayList<T>()
        for (element in target) {
            if (predicate.apply(element)) {
                result.add(element)
            }
        }
        return result
    }
}
